### PR TITLE
Support getattr for tensor subclasses in pre-dispatch export via patching tensor.getattr

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143946

Previous discussion: https://github.com/pytorch/pytorch/pull/143671#issuecomment-2560112499 and https://github.com/pytorch/pytorch/pull/143671

Differential Revision: [D67693609](https://our.internmc.facebook.com/intern/diff/D67693609)

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67693609).

File: torch/export/_trace.py [hunk 1]
```diff
@@ -1472,7 +1480,93 @@ def wrapped_fn(*args):
...
+                        )
+                        for mode in torch_function_mode_stack:
+                            if isinstance(mode, PreDispatchTorchFunctionMode):
+                                tracer = mode.tracer
+                                proxy = get_proxy_slot(self, tracer).proxy
+                                inner_proxy = tracer.create_proxy(
+                                    "call_function", builtins.getattr, (proxy, attr), {}
+                                )
+                                # Even though the name says tensor tracking, the impl
+                                # suggests it tracks everything?
```
Comment (bdhirsh) [inline torch/export/_trace.py:1511]: I think in practice it tracks tensors, and also SymInts (once SymInt support was added after-the-fact)

File: torch/export/_trace.py [hunk 2]
```diff
@@ -1472,7 +1480,93 @@ def wrapped_fn(*args):
...
+                Context manager that temporarily monkey patches
+                tensor.__getattribute__ so that we can intercept it at
+                torch_function layer.
+                """
+
+                # Dictionary that tracks subclass type to original getattr function
+                # and the attributes we can proxy.
+                tensor_type_to_old_getattribute = {}
+                for arg in args:
+                    if is_traceable_wrapper_subclass(arg):
```
Comment (bdhirsh) [inline torch/export/_trace.py:1529]: nit: you can short-circuit and also skip if `type(arg)` has already been seen (say you have 100 params that are all of the same subclass type)

File: torch/export/_trace.py [hunk 3]
```diff
@@ -1472,7 +1480,93 @@ def wrapped_fn(*args):
...
+                        attrs_to_proxy = []
+                        for attr in subclass_attrs:
+                            val = getattr(arg, attr)
+                            # If it is a static function or method, we should always inline
+                            if not isinstance(
+                                val, (types.FunctionType, types.MethodType)
+                            ):
+                                attrs_to_proxy.append(attr)
+                        inner_tensors, _ = arg.__tensor_flatten__()
+                        attrs_to_proxy.extend(inner_tensors)
```
Comment (bdhirsh) [inline torch/export/_trace.py:1541]: This line seems redundant? Any attribute names of "inner_tensors" that are returned from __tensor_flatten__ should already show up as extra attributes that you found from `dir(arg)`.

File: torch/export/_trace.py [hunk 4]
```diff
@@ -1472,7 +1480,93 @@ def wrapped_fn(*args):
...
+                        tensor_type_to_old_getattribute[arg.__class__] = (
+                            arg.__class__.__getattribute__,
+                            attrs_to_proxy,
+                        )
+
+                try:
+                    for arg in args:
```
Comment (bdhirsh) [inline torch/export/_trace.py:1548]: Any reason you're looping through `args` a second time here? I think you can just loop through `tensor_type_to_old_getattribute.items()` (the key is the tensor type, which should be equivalent to `arg.__class__`

File: test/export/test_export.py [hunk 5]
```diff
@@ -1672,6 +1672,94 @@ def forward(self, x):
...
+                self.p1 = torch.nn.Parameter(torch.ones(3, 4))
+                self.p2 = torch.nn.Parameter(
+                    ConstantExtraMetadataTensor(
+                        ConstantExtraMetadataTensor(torch.ones(3, 4)),
+                    )
+                )
+
+            def forward(self, x):
+                res = 2 * self.p1 + self.p2
+                res2 = res + res.constant_attribute
```
Comment (bdhirsh) [inline test/export/test_export.py:1738]: One thing to call out is that if you look at the expecttest result below, there isn't a node for the `res.constant_attribute` access, e.g.:
```
%constant_attribute = call_function[target=builtins.getattr](args = (res, "constant_attribute"), kwargs = {})
%add_1 = call_function[target=torch.ops.aten.add.Tensor](args = (%add, %constant_attribute), kwargs = {})
```

Instead, the constant gets inlined in the graph to its value (4):
```
%add_1  = call_function[target=torch.ops.aten.add.Tensor](args = (%add, 4), kwargs = {})
```

This... may or may not be ok. It seems fine for this example, where the constant is just an integer. Once we get constructor support working, though, then you'll have to make sure code like this can be traced properly:
```
def f(subclass):
    constant_metadata_namedtuple = subclass.metadata
    new_subclass = Subclass(subclass.elem, constant_metadata_namedtuple)
```

Where if the constant metadata is something a bit more complicated (e.g. a named tuple holding constant values, which is the case for float8), then we'll need to make sure it shows up in a reasonable way as a contructor argument.

File: test/export/test_export.py [hunk 6]
```diff
@@ -1672,6 +1672,94 @@ def forward(self, x):
...
+        from torch.testing._internal.two_tensor import TwoTensor
+
+        class Foo(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.p1 = torch.nn.Parameter(torch.ones(3, 4))
+                self.p2 = torch.nn.Parameter(
+                    TwoTensor(
+                        TwoTensor(torch.ones(3, 4), torch.ones(3, 4)),
+                        TwoTensor(torch.ones(3, 4), torch.ones(3, 4)),
```
Comment (bdhirsh) [inline test/export/test_export.py:1689]: A few more testing ideas:

(1) a test where you have `TwoTensor(ConstantExtraMetadataTensor, ConstantExtraMetadataTensor)`, to test that we are properly proxying getattrs on subclass types that don't show up in the top-level arguments

(2) a test with a subclass that has more complicated constant metadata (e.g. a NamedTuple, to be a bit more representative of [Float8](https://github.com/pytorch/ao/blob/main/torchao/float8/float8_tensor.py#L46)). Mostly to make sure that something reasonable happens today, and also that we can augment later when we get constructor support (see comment below)

File: torch/export/_trace.py [hunk 7]
```diff
@@ -1472,7 +1480,93 @@ def wrapped_fn(*args):
...
+                out = original_getattr(self, attr)
+                if attr in attrs_to_proxy:
+                    if torch._C._is_torch_function_mode_enabled():
+                        # When we get here there is no guarantee that we will hit the
+                        # PreDispatchTorchFunctionMode, so we manually peak into the torch
+                        # function mode list and tweak the PreDispatchTorchFunctionMode.
+                        # This has side effect of proxying stuff like
+                        # proxy.node.meta["val"] = extract_val(val) because at that time, torch function
+                        # mode is still active. It seems bad to turn it off inside proxy_tensor.py, so
+                        # I guess we will just rely on DCE for now to remove extra stuff like detach
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply torch/export/_trace.py:1499]: Yup! extract_val calls detach and we are proxying that in the graph.

File: torch/export/_trace.py (hunk 4 above)
Comment (tugsbayasgalan) [reply torch/export/_trace.py:1548]: Ah yep true!

File: torch/export/_trace.py [hunk 8]
```diff
@@ -1510,6 +1604,8 @@ def _is_impure(node):
                     torch.ops.profiler._record_function_enter.default,
                     torch.ops.profiler._record_function_enter_new.default,
                     torch.ops.profiler._record_function_exit._RecordFunction,
+                    torch.ops.aten.detach.default,
+                    builtins.getattr,
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply torch/export/_trace.py:1608]: I think this shows up when you are calling extract_val because then you access all inner attributes and check if they are FakeTensor etc.

File: torch/export/_trace.py (hunk 3 above)
Comment (tugsbayasgalan) [reply torch/export/_trace.py:1541]: Yep good point!

File: torch/export/_trace.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/export/_trace.py]: got it. the extra detach() that gets DCEd sounds pretty harmless.

If we decide for some reason we want to avoid it later then we can always roll our own `track_tensor_tree` that just manually calls `aten.detach.default(x)` instead of `x.detach()`

File: test/export/test_export.py (hunk 6 above)
Comment (tugsbayasgalan) [reply test/export/test_export.py:1688]: Discussed offline. (2) won't actually be tested comprehensively given current state (need an operator or subclass constructor that consumes complicated metadata).

File: torch/export/_trace.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/export/_trace.py:1601]: I don't follow though - why do we need to mark these two functions as impure?

My understanding is the main reason we mark some ops as impure is that they has some nonlocal side-effects, and we want to make sure that our graph passes like DCE don't remove them just because they look like dead code.

But both detach and getattr are actual pure functions (they don't have secret side effects)

File: torch/_functorch/_aot_autograd/schemas.py [hunk 9]
```diff
@@ -310,13 +310,6 @@ def __post_init__(self):
         # sanity assert to make sure we don't leak memory
         assert is_fake(self.original_subclass)
 
-        # This saves the type of subclass nested structure to compare
-        # against runtime tangent inputs. We do wanna compute this at AOT
-        # time as it is invoked in hot-path
-        from .subclass_utils import get_types_for_subclass
-
-        self.subclass_type = get_types_for_subclass(self.original_subclass)
```
Comment (tugsbayasgalan) [inline torch/_functorch/_aot_autograd/schemas.py:318]: Don't know why this still exist, i deleted it and repurposed the helper function for my use case lol

File: test/export/test_export.py (hunk 6 above)
Comment (bdhirsh) [reply test/export/test_export.py:1688]: Tbh it still seems worth testing to make sure something reasonable happens. e.g.:

```
named_tuple_val = subclass_tensor.metadata
# `named_tuple_val.foo` should get inlined as a constant
return subclass_tensor.elem * named_tuple_val.foo
```

File: torch/export/_trace.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply torch/export/_trace.py:1601]: Hmmm so fx DCE calls is_impure function on every node before DCE-ing. These special cased operators will return False (suggesting it is safe to delete). Since we don't want to actually run DCE aggresively, we will by default return True. As a result, only nodes that are marked in this if statement will be deleted.

[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/export/_trace.py:1601]: oh I see... it's probably worth a comment for every op that we add to this list (otherwise a year from now someone reading this will have no idea why getattr/detach need to be DCE'd but not other ops). Can you add one describing the situation? (and in particular - that it's not fundamental, and that we can avoid tracing these "dead" getattr/detach calls in principle, but it would require carefully re-writing `track_tensor_tree` to not do any tensor method calls)

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67693609).

File: test/export/test_export.py (hunk 6 above)
Comment (tugsbayasgalan) [reply test/export/test_export.py:1688]: Updated the test case.

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67693609).

File: torch/_functorch/_aot_autograd/subclass_utils.py [hunk 10]
```diff
@@ -54,15 +55,19 @@ def maybe_suggest_memory_format(
...
-def get_types_for_subclass(tensor_subclass):
-    if not is_traceable_wrapper_subclass(tensor_subclass):
-        return ["Tensor"]
-    inner_keys, _ = tensor_subclass.__tensor_flatten__()
-    result = []
-    for key in inner_keys:
-        inner_tensor = getattr(tensor_subclass, key)
-        result.extend(get_types_for_subclass(inner_tensor))
-    return result
+def get_types_for_subclass(tensor_subclass) -> Dict[Any, List[Any]]:
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/subclass_utils.py:58]: can we do better than Any -> Any? It seems like this is actually `Dict[type, List[torch.Tensor]]`?

From the function name it's not immediately clear why we're returning a dict - maybe worth a quick comment on the top saying that the values are all of the instances that we found of each subclass type

File: torch/export/_trace.py [hunk 11]
```diff
@@ -1472,7 +1480,86 @@ def wrapped_fn(*args):
...
+                                    tracer = mode.tracer
+                                    proxy = get_proxy_slot(self, tracer).proxy
+                                    inner_proxy = tracer.create_proxy(
+                                        "call_function",
+                                        builtins.getattr,
+                                        (proxy, attr),
+                                        {},
+                                    )
+                                    # Even though the name says tensor tracking, the impl
+                                    # suggests it tracks everything?
```
Comment (bdhirsh) [inline torch/export/_trace.py:1516]: You probably want to remove the comment before merging. But from what we talked about in the test, this function will only actually track Tensors and SymInts (for everything else, we it will not track any proxies because other outputs will be treated as constants during tracing)

**Review by bdhirsh — APPROVED** [review]
left a few more comments, but overall lgtm!

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67693609).
